django_app/redbox_app/redbox_core/models.py

```diff
@@ -0,0 +1,923 @@
+    return string.replace("\x00", "\ufffd") if string else string
+
+
+class ChatLLMBackend(models.Model):
+    """https://python.langchain.com/docs/how_to/chat_models_universal_init/"""
+
+    class Providers(models.TextChoices):
+        OPENAI = "openai"
+        ANTHROPIC = "anthropic"
+        AZURE_OPENAI = "azure_openai"
+        GOOGLE_VERTEXAI = "google_vertexai"
+        GOOGLE_GENAI = "google_genai"
+        BEDROCK = "bedrock"
+        BEDROCK_CONVERSE = "bedrock_converse"
+        COHERE = "cohere"
+        FIREWORKS = "fireworks"
+        TOGETHER = "together"
+        MISTRALAI = "mistralai"
+        HUGGINGFACE = "huggingface"
+        GROQ = "groq"
+        OLLAMA = "ollama"
+
+    name = models.CharField(
+        max_length=128,
+        help_text="The name of the model, e.g. “gpt-4o”, “claude-3-opus-20240229”.",
+    )
+    provider = models.CharField(max_length=128, choices=Providers, help_text="The model provider")
+    description = models.TextField(null=True, blank=True, help_text="brief description of the model")
+    is_default = models.BooleanField(default=False, help_text="is this the default llm to use.")
+    enabled = models.BooleanField(default=True, help_text="is this model enabled.")
+    display = models.CharField(max_length=128, null=True, blank=True, help_text="name to display in UI.")
+
+    class Meta:
+        constraints = [UniqueConstraint(fields=["name", "provider"], name="unique_name_provider")]
+
+    def __str__(self):
+        return self.display or self.name
+
+    def save(self, *args, **kwargs):
+        if self.is_default:
+            ChatLLMBackend.objects.filter(is_default=True).update(is_default=False)
+        super().save(*args, **kwargs)
+
+
+class AbstractAISettings(models.Model):
+    chat_backend = models.ForeignKey(ChatLLMBackend, on_delete=models.CASCADE, help_text="LLM to use in chat")
+    temperature = models.FloatField(default=0, help_text="temperature for LLM")
+
+    class Meta:
+        abstract = True
+
+    def save(self, *args, **kwargs):
+        if not self.chat_backend_id:
+            self.chat_backend = ChatLLMBackend.objects.get(is_default=True)
+        return super().save(*args, **kwargs)
+
+
+class AISettings(UUIDPrimaryKeyBase, TimeStampedModel, AbstractAISettings):
+    label = models.CharField(max_length=50, unique=True)
+
+    # LLM settings
+    context_window_size = models.PositiveIntegerField(null=True, blank=True)
+    llm_max_tokens = models.PositiveIntegerField(null=True, blank=True)
+
+    # Prompts and LangGraph settings
+    max_document_tokens = models.PositiveIntegerField(null=True, blank=True)
+    self_route_enabled = models.BooleanField(null=True, blank=True)
+    map_max_concurrency = models.PositiveIntegerField(null=True, blank=True)
+    stuff_chunk_context_ratio = models.FloatField(null=True, blank=True)
+    recursion_limit = models.PositiveIntegerField(null=True, blank=True)
+
+    chat_system_prompt = models.TextField(null=True, blank=True)
```

Comment:
Am I right in thinking this is all needed because it was weirdly picking up empty strings and using them for the prompts? 

I wonder if this is something we could improve while refactoring? Especially as we are slimming down the number of prompts.

Reply:
we added this to enable us to test different system prompts without having to redeploy the code. in practice it hasn't been very useful and could be removed